pkgsStatic.stdenv.cc.cc: put static libs in $lib

alyssais opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This fixes static C++ builds with libtool.  Libtool requires its archive files to be in the same directory as the library files it's using.  So for dynamic builds, the .la files need to be in the same directory as the .so files, and for static builds, they need to be in the same directory as the .a files.  Since shared libraries are disabled in static builds, the .la files won't be moved to the lib output, and libtool will be able to find libstdc++.a.

It's not 100% required to stop building the shared libraries when doing static builds — we'd just have to disable the output.  But they'd bloat the output, and I doubt their useful anyway for reasons explained in the commit message.

Fixes: https://github.com/NixOS/nixpkgs/issues/76873
Fixes: https://github.com/NixOS/nixpkgs/issues/108534

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

alyssais committed:
Revert "gcc: Always pass `--enable-shared` by default"

This reverts commit 8e482321805542e08eaf8819a6ad94ea4ba20370.

Since pkgsStatic.stdenv.cc can only produce static binaries, there's
no reason to include that compilers e.g. libstdc++.so.

Ericson2314 commented:
What about just moving the `.a` files in the static build? I generally prefer if we always have the same outputs.

alyssais commented:
@Ericson2314 hmm, I'm not sure I like the semantics of the output changing.  Normally, the lib output is for libraries that dependents will link against, but this would change it (in the static case) to being stuff needed to use the compiler, i.e. the exact same case as the default output.  It would never make sense to use one output without the other.

Ericson2314 commented:
Oh yes I agree, but why can't we just put the out and lib libraries in the same outputs they are in for the regular build with the static build?

alyssais commented:
@Ericson2314 and I discussed this a bunch [on Matrix](https://logs.nixos.dev/room/!ayCRiZriCVtuCUpeLp:nixos.org/?anchor=$Cb0TwRWv1jQoDt05f39pMwZ0QC1HiAfa16lYU82-7rA&offset=240).

alyssais committed:
pkgsStatic.stdenv.cc.cc: put static libs in $lib

libtool requires the library files its .la files describe to be in the
same directory as those files.  We only do this for compilers without
shared libraries, so that the libraries in $lib are always the
libraries the package set is supposed to use.

Fixes: https://github.com/NixOS/nixpkgs/issues/76873
Fixes: https://github.com/NixOS/nixpkgs/issues/108534

alyssais commented:
@Ericson2314 I've changed this to put static libraries in $lib when the target platform is static, as we discussed.  I was a bit skeptical this would just work, but I've built all of Spectrum (which a very heavy `pkgsStatic` user), and everything seems fine, plus it still fixes the protobuf issue.

Ericson2314 commented:
@alyssais Great! I think this needs `staging`? But otherwise it looks good to me!

alyssais commented:
> I think this needs `staging`?

You sure?  It's less than 1K rebuilds.

alyssais commented:
@Ericson2314 changed.

tobim commented:
The first commit of this PR broke `pkgsStatic.python3Minimal`.
It fails with
```
/nix/store/4rvj9yvp5wqlq6ps55mxi5x6i8h477cb-x86_64-unknown-linux-musl-binutils-2.35.2/bin/x86_64-unknown-linux-musl-ld: cannot find -lgcc_eh
collect2: error: ld returned 1 exit status
```
I quick search pointed me towards https://libc-help.sourceware.narkive.com/gQzN0OOU/why-does-static-gnulib-include-lgcc-eh#
which is old, but at least does explain the error.
I'm not sure if this worked by accident before.

alyssais commented:
It also broke `pkgsStatic.libunwind`. :(  I think we've discovered the reason the shared libraries were built unconditionally in the first place.

tobim commented:
`libundwind` gets patched in nixpkgs:
```
  postPatch = if stdenv.cc.isClang then ''
    substituteInPlace configure.ac --replace "-lgcc_s" ""
  '' else lib.optionalString stdenv.hostPlatform.isMusl ''
    substituteInPlace configure.ac --replace "-lgcc_s" "-lgcc_eh"
  '';
```
We just need to add an extra case for `isStatic`.
I already fixed python in #174285.

alyssais commented:
I've done some research, and it seems like when GCC is built with `--disable-shared` everything just ends up in libgcc.a, rather than being split between libgcc_s.so, libgcc.a, and libgcc_eh.a.  Interestingly, if I just change libunwind to always use libgcc rather than libgcc_eh, the non-static `pkgsMusl.libunwind` still builds.  So maybe we can just always use libgcc instead of libgcc_eh?  I need to do a bit more experimentation.

I'm also confused that you were able to fix `pkgsStatic.python3Minimal` by just removing the linker flag all together.  I'm going to check whether it's still needed in the other cases / why it was added in the first place.

Ericson2314 commented:
I am confused too, but I like that is PR seems to be pushing us towards solutions of *deleting* code / at least making it less conditional :)

tobim commented:
I blame python's build scaffold for that, it fails recognize that it should to link against `libgcc_s` and `libgcc_eh` if they exist. They don't in `pkgsStatic` land, so we don't have to add the override there.

tobim commented:
#177934 for libunwind.

alyssais commented:
On Thu, Jun 16, 2022 at 02:26:07PM -0700, tobim wrote:
> I blame python's build scaffold for that, it fails recognize that it should to link against `libgcc_s` and `libgcc_eh` if they exist. They don't in `pkgsStatic` land, so we don't have to add the override there.

This can't be the whole story, because if I comment out the Musl line,
pkgsMusl.python3Minimal still builds despite the presence of libgcc_eh
in that case.
